### PR TITLE
Update getTenantInstances call for controller and separate POST operations on it

### PR DESCRIPTION
**Pull request by tibrewalpratik17**
labels:
   1. `bugfix`
   2. `backward-incompatible`

We found an issue with `listInstanceOrToggleTenantState` endpoint today in our production.

- We do a POST operation in a API which is marked as GET. This leads to confusion and treats this API as mainly a NOOP.
- We don't check if the state passed is a valid state in the list of enums - ENABLE, DISABLE, DROP.
- If the state is not valid, we end-up disabling all the instances. 

This led to ingestion and query failures in our production cluster.

This PR separates the POST operation in a separate POST API. Adds a check for matching the state passed in the API parameter. Removes DROP operation as this is covered by deleteTenant API. Disables instances only when `DISABLE` is passed as the State.

### CODE REVIEW

**Comment by tibrewalpratik17** [comment]
cc @chenboat

File: pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotTenantRestletResource.java [hunk 1]
```diff
@@ -203,19 +204,36 @@ public TenantsList getAllTenants(
...
+      @ApiParam(value = "Tenant name", required = true) @PathParam("tenantName") String tenantName,
+      @ApiParam(value = "Tenant type (server|broker)") @QueryParam("type") String tenantType,
+      @ApiParam(value = "Table type (offline|realtime)") @QueryParam("tableType") String tableType) {
+    return listInstancesForTenant(tenantName, tenantType, tableType);
+  }
+
+  @POST
+  @Path("/tenants/{tenantName}")
+  @Produces(MediaType.APPLICATION_JSON)
+  @ApiOperation(value = "enable/disable/drop a tenant")
```
Comment (Jackie-Jiang) [inline pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotTenantRestletResource.java:222]: Suggest only allowing enable and disable. Drop should be achieved by a DELETE request

File: pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotTenantRestletResource.java [hunk 2]
```diff
@@ -203,19 +204,36 @@ public TenantsList getAllTenants(
...
+  @ApiResponses(value = {
+      @ApiResponse(code = 200, message = "Success"),
+      @ApiResponse(code = 500, message = "Error applying state to tenant")
+  })
+  public SuccessResponse toggleTenantState(
```
Comment (Jackie-Jiang) [inline pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotTenantRestletResource.java:227]: Seems we also have `changeTenantState()` for the same purpose. Can you check if they are doing the exact same thing? If so, we should remove the duplicate logic, and mark `changeTenantState()` as deprecated

File: pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotTenantRestletResource.java [hunk 3]
```diff
@@ -284,19 +302,24 @@ private String toggleTenantState(String tenantName, String stateStr, @Nullable S
...
+
+    if (StateType.DISABLE.name().equalsIgnoreCase(stateStr)) {
+      for (String instance : allInstances) {
         instanceResult.put(instance, JsonUtils.objectToJsonNode(_pinotHelixResourceManager.disableInstance(instance)));
       }
+      return new SuccessResponse("Disabled tenant " + tenantName + " successfully.");
     }
 
-    return null;
+    return new SuccessResponse("No-Op done on tenant " + tenantName);
```
Comment (Jackie-Jiang) [inline pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotTenantRestletResource.java:322]: Will we ever reach this branch?

File: pinot-spi/src/main/java/org/apache/pinot/spi/utils/builder/ControllerRequestURLBuilder.java [hunk 4]
```diff
@@ -163,6 +163,12 @@ public String forBrokerTablesGet(String state) {
     return StringUtil.join("/", _baseUrl, "brokers", "tables", "?state=" + state);
   }
 
+  public String forTenantInstancesToggle(String tenant, String tenantType, String state) {
+    StringBuilder url = new StringBuilder(StringUtil.join("/", _baseUrl, "tenants", tenant));
+    url.append("?type=").append(tenantType);
+    url.append("&state=").append(state);
+    return String.valueOf(url);
+  }
```
Comment (Jackie-Jiang) [inline pinot-spi/src/main/java/org/apache/pinot/spi/utils/builder/ControllerRequestURLBuilder.java:171]: (nit) empty line below

**Review by Jackie-Jiang — COMMENTED** [review]
This is great. We definitely shouldn't use GET to change or even drop tenant

File: pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotTenantRestletResource.java (hunk 1 above)
Comment (tibrewalpratik17) [reply pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotTenantRestletResource.java]: I see there is already a delete tenant API so removing it from here.

File: pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotTenantRestletResource.java (hunk 2 above)
Comment (tibrewalpratik17) [reply pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotTenantRestletResource.java]: Yes looks like the logic is same. Marked the other one as deprecated.

File: pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotTenantRestletResource.java (hunk 3 above)
Comment (tibrewalpratik17) [reply pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotTenantRestletResource.java]: Removed.

File: pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotTenantRestletResource.java [hunk 5]
```diff
@@ -276,27 +292,17 @@ private String toggleTenantState(String tenantName, String stateStr, @Nullable S
...
+    if (StateType.ENABLE.name().equalsIgnoreCase(stateStr)) {
+      for (String instance : allInstances) {
         instanceResult.put(instance, JsonUtils.objectToJsonNode(_pinotHelixResourceManager.enableInstance(instance)));
-      } else {
-        instanceResult.put(instance, JsonUtils.objectToJsonNode(_pinotHelixResourceManager.disableInstance(instance)));
       }
     }
-
-    return null;
+    return new SuccessResponse("Changed state of tenant " + tenantName + " to " + stateStr + " successfully.");
```
Comment (chenboat) [inline pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotTenantRestletResource.java:305]: There are three states: can we add the from state to the response?

File: pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotTenantRestletResource.java [hunk 6]
```diff
@@ -260,7 +276,7 @@ private String getTablesServedFromTenant(String tenantName) {
     return resourceGetRet.toString();
   }
 
-  private String toggleTenantState(String tenantName, String stateStr, @Nullable String tenantType) {
+  private SuccessResponse toggleTenantState(String tenantName, String stateStr, @Nullable String tenantType) {
```
Comment (chenboat) [inline pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotTenantRestletResource.java:279]: toggle means switch between 2 states. But the StateType enum has 3 states. We should use changeTenantState to reduce the confusion -- unless you can explain in javadoc why some state is not used.

Comment (tibrewalpratik17) [reply pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotTenantRestletResource.java:279]: Now we are only using 2 states here - enable / disable. So toggle still makes sense. DROP is taken care of deleteTenant API which was already there.

File: pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotTenantRestletResource.java (hunk 5 above)
Comment (tibrewalpratik17) [reply pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotTenantRestletResource.java:305]: We are using 2 states in this API now just enable/disable.

**Review by Jackie-Jiang — APPROVED** [review]

**Pull request merged**